### PR TITLE
Closes #3717 compatibility with wpx hosting

### LINKED ISSUE #3717: Automatic compatibility with WPX hosting and their CDN
**Is your feature request related to a problem? Please describe.**
WP Rocket cache-control header for text/html with max-age=0 is not compatible with WPX custom-built CDN.

**Describe the solution you'd like**
Automatically remove the header for text/html for websites hosted with WPX.

**Additional context**
Ticket: https://secure.helpscout.net/conversation/1464674180/250829/

Currently, the header can be removed with our helper plugin: 
https://github.com/wp-media/wp-rocket-helpers/tree/master/htaccess/wp-rocket-htaccess-no-expires-html

### PR DESCRIPTION
**Pull request by jeawhanlee**
## Description

Adds compatibility with WPX Cloud hosting.

Fixes #3717 

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?
No

## How Has This Been Tested?
Automated tests.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by CrochetFeve0251 — APPROVED** [review]

File: tests/Integration/inc/ThirdParty/Hostings/WPXCloud/varnishIP.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<?php
+namespace WP_Rocket\Tests\Integration\inc\ThirdParty\Hostings\WPXCloud;
+
+use WP_Rocket\ThirdParty\Hostings\WPXCloud;
+use WPMedia\PHPUnit\Integration\TestCase;
```
Comment (remyperona) [inline tests/Integration/inc/ThirdParty/Hostings/WPXCloud/varnishIP.php]:
```suggestion
use WP_Rocket\Tests\Integration\TestCase;
```

File: tests/Integration/inc/ThirdParty/Hostings/WPXCloud/varnishIP.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * @covers \WP_Rocket\ThirdParty\Hostings\WPXCloud::varnish_ip
+ * @group WPXCloud
+ */
+class Test_VarnishIP extends TestCase {
+
+	/**
+	 * @dataProvider providerTestData
```
Comment (remyperona) [inline tests/Integration/inc/ThirdParty/Hostings/WPXCloud/varnishIP.php]:
```suggestion
	 * @dataProvider configTestData
```

File: tests/Integration/inc/ThirdParty/Hostings/WPXCloud/varnishIP.php [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+	public function testShouldReturnWPXCloudVarnishIP( $varnish_ip, $expected ) {
+		$this->assertSame(
+			$expected,
+			apply_filters( 'rocket_varnish_ip', $varnish_ip )
+		);
+	}
+
+	public function providerTestData() {
+		return $this->getTestData( __DIR__, 'varnishIP' );
+	}
```
Comment (remyperona) [inline tests/Integration/inc/ThirdParty/Hostings/WPXCloud/varnishIP.php]:
```suggestion
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by vmanthos** [comment]
@jeawhanlee After discussing this with Piotr, the Varnish add-on should be automatically enabled when we detect the site is hosted on WPX. We are doing this already for Cloudways:
https://github.com/wp-media/wp-rocket/blob/develop/inc/ThirdParty/Hostings/Cloudways.php

Also, while purging the whole cache works fine, partial and homepage cache clearing isn't working.

I logged the `$url` and `$args` we use here:
https://github.com/wp-media/wp-rocket/blob/14eab568207523ee0207a0e09f07d48e99db4edf/inc/Addon/Varnish/Varnish.php#L94

```PHP
[ 2022-11-01 05:45:20 ] Purge URL: http://127.0.0.1:6081/hello-world/.*
Array
(
    [method] => PURGE
    [blocking] => false
    [redirection] => 0
    [headers] => Array
        (
            [host] => wpxnew.rocketlabsqa.ovh
            [X-Purge-Method] => regex
        )

)

[ 2022-11-01 05:45:20 ] Purge URL: http://127.0.0.1:6081/
Array
(
    [method] => PURGE
    [blocking] => false
    [redirection] => 0
    [headers] => Array
        (
            [host] => wpxnew.rocketlabsqa.ovh
            [X-Purge-Method] => default
        )

)
```

I tried different combinations of scheme and regex but the Varnish cache wasn't purged. 
@piotrbak you may need to reach out to WPX and check if there is anything wrong with the requests we sent.

**Comment by alxtv** [comment]
Alex from WPX here:

@vmanthos The issue with purging homepage and partial cache has been resolved. 

Can you please try again ?

**Comment by vmanthos** [comment]
Thank you @alxtv!

Indeed, both partial cache clearing and homepage cache clearing is working as expected. 🎉

**Comment by vmanthos** [comment]
@alxtv WP Rocket's preload feature isn't working on our test site on wpx.net. 

We are using `WP Rocket/Preload` as the user agent and requests from that have an `error 403`:
```
45.77.227.117 5.254.55.39 - [01/Nov/2022:11:37:19 +0000] "GET /hello-world/ HTTP/1.1" 403 1074 "https://wpxnew.rocketlabsqa.ovh/hello-world/" "WP Rocket/Preload" "-" 0 
45.77.227.117 5.254.55.39 - [01/Nov/2022:11:37:19 +0000] "GET /will-it-purge/ HTTP/1.1" 403 1074 "https://wpxnew.rocketlabsqa.ovh/will-it-purge/" "WP Rocket/Preload" "-" 0 
45.77.227.117 5.254.55.39 - [01/Nov/2022:11:37:20 +0000] "GET /sample-page/ HTTP/1.1" 403 1074 "https://wpxnew.rocketlabsqa.ovh/sample-page/" "WP Rocket/Preload" "-" 0 
45.77.227.117 5.254.55.39 - [01/Nov/2022:11:37:20 +0000] "GET /category/uncategorized/ HTTP/1.1" 403 1074 "https://wpxnew.rocketlabsqa.ovh/category/uncategorized/" "WP Rocket/Preload" "-" 0 
```

Would you mind taking a looking? If I have to reach out to someone else about this, please let me know who that is. 🙏

**Update:**

In the error log, there are entries like the following coming from `ModSecurity`:
```
 ModSecurity: Access denied with code 403, [Rule: 'REQUEST_HEADERS:User-Agent' '@contains Preload'] [id "9670013"] [msg "malicious useragent Preload"]
```

**Comment by alxtv** [comment]
@vmanthos We have "disabled" the preload feature by default a long time ago as it caused high CPU usage for many websites (websites with thousands auto-generated posts/pages, sites using some unoptimized/bad-coded plugins/themes, sites with many products or big databases and etc).
We are enabling it on request via support ticket or livechat.

I've enabled it for your website so you can check if everything works fine.

I see that the description in error_log is not accurate so I've updated it too.

**Comment by vmanthos** [comment]
Thank you very much for the quick response and for the explanation, @alxtv!

Preloading works now.

Maybe we can work together (pinging @piotrbak) to investigate cases like the one you described so you don't have to disable preload for our common customers.

**Comment by alxtv** [comment]
Sure, we can discuss this via email.

We're open to work on this and find a better solution for the cache preload feature.

We are planning to do new round of performance/speed tests and comparisons once these changes are included in the plugin so additionally I will include the Preload feature in tests and see how it works now with different setups.
Will provide you with feedback once we have some results.

**Review by remyperona — APPROVED** [review]

**Review by vmanthos — APPROVED** [review]
This is working as expected. 👍

**Pull request merged**